### PR TITLE
Implement ICorProfilerInfo13 to support object handle creation/destruction/access

### PR DESCRIPTION
**Pull request by chrisnas**
In the context of https://github.com/dotnet/runtime/issues/49424, 3 methods are added to ICorProfilerInfo to manage strong, weak and pinned handles. 

The first usage will be to wrap the reference provided by AllocationTick_V4 by a weak reference handle and use them to figure out if the object is still alive after a certain amount of time. This should help building automatic memory leak detection system.

Note: the corresponding documentation is made available via https://github.com/dotnet/docs/pull/30131

### CODE REVIEW

**Comment by teo-tsirpanis** [comment]
@davmason

**Comment by davmason** [comment]
@chrisnas  if closing it was by accident, you should be able to reopen it. I can't for some reason

**Comment by chrisnas** [comment]
git error...

File: src/tests/profiler/native/README.md
Comment (davmason) [inline src/tests/profiler/native/README.md:50]: Thanks for updating the docs!

File: src/tests/profiler/native/handlesprofiler/handlesprofiler.cpp
Comment (davmason) [inline src/tests/profiler/native/handlesprofiler/handlesprofiler.cpp:107]: Nit: here and line 121 please put the else and if on the same line

File: src/tests/profiler/native/handlesprofiler/handlesprofiler.cpp
Comment (davmason) [inline src/tests/profiler/native/handlesprofiler/handlesprofiler.cpp:255]: It is not supported to get objects from a destroyed handle 😄. When you destroy the handle it is returned to the pool of available handles and you would be potentially getting some random other object, or garbage.

If you wanted to verify that the object is indeed collected you could use a second weak handle and verify that it returns NULL. I don't think it's necessary to add that though.

File: src/tests/profiler/native/handlesprofiler/handlesprofiler.cpp
Comment (davmason) [inline src/tests/profiler/native/handlesprofiler/handlesprofiler.cpp:211]: It would be good to add a check that the pinned handle never changes. Either checking the generation is stable or that the ObjectId returned never changes should suffice

File: src/tests/profiler/handles/handles.cs
Comment (davmason) [inline src/tests/profiler/handles/handles.cs:96]: It's not clear to me if you intend to only collect certain generations, the code makes it seem like you expect the argument to `GC.Collect` to be the count of the GCs so far. `GC.Collect(0)` will collect Gen 0, `GC.Collect(1)` will collect Gen 0 and Gen 1, etc.

**Review by davmason — CHANGES_REQUESTED** [review]
Just a few small changes, otherwise looks good to me!

File: src/coreclr/vm/proftoeeinterfaceimpl.cpp
Comment (chrisnas) [reply src/coreclr/vm/proftoeeinterfaceimpl.cpp:7483]: ✅ Done

File: src/coreclr/vm/proftoeeinterfaceimpl.cpp
Comment (chrisnas) [reply src/coreclr/vm/proftoeeinterfaceimpl.cpp:7510]: ✅ Good point. I'm also setting *pHandle to NULL

File: src/coreclr/vm/proftoeeinterfaceimpl.cpp
Comment (chrisnas) [reply src/coreclr/vm/proftoeeinterfaceimpl.cpp:7475]: Make sense for DestroyHandle and GetObjectIDFromHandle.
However, it seems mandatory for CreateHandle: it should ensure that the given ObjectID "really" refers to an existing instance at that particular address (i.e. received synchronously from a callback without the risk of a GC moving objects around). For example, getting an instance asynchronously from AllocationTick_V4 via EventPipe might have this kind of issue.

File: src/tests/profiler/native/README.md
Comment (chrisnas) [reply src/tests/profiler/native/README.md:50]: > Thanks for updating the docs!

I'm starting to look at the MS Docs too  ;^)

File: src/tests/profiler/native/handlesprofiler/handlesprofiler.cpp
Comment (chrisnas) [reply src/tests/profiler/native/handlesprofiler/handlesprofiler.cpp:255]: ✅  Thanks for the details: I'm removing the comment and updating the test accordingly

File: src/tests/profiler/handles/handles.cs
Comment (chrisnas) [reply src/tests/profiler/handles/handles.cs:96]: ✅ This is definitively misleading  :^(
 I imagined that I would use the collected generation as an additional parameter but I did not used it. Updating the code to only collect gen0.

**Comment by davmason** [comment]
@chrisnas I think this is ready to merge, when are you planning on doing the docs on docs.microsoft.com? I would really prefer to wait for a docs PR before merging, otherwise docs tend to never get done.

**Comment by chrisnas** [comment]
> @chrisnas I think this is ready to merge, when are you planning on doing the docs on docs.microsoft.com? I would really prefer to wait for a docs PR before merging, otherwise docs tend to never get done.

@davmason : is there anything else to do beyond https://github.com/dotnet/docs/pull/30131?

**Comment by davmason** [comment]
> is there anything else to do beyond https://github.com/dotnet/docs/pull/30131?

Nope, just missed that PR. Thanks, I'll merge this now

**Review by davmason — APPROVED** [review]
LGTM

**Pull request merged**
